File: java/15. 3Sum.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {
      List<List<Integer>> result = new ArrayList<>();

      Arrays.sort(nums);

      for(int i = 0; i < nums.length - 1; i++) {
          int currentElement = nums[i];

          int l = i + 1;
          int r = nums.length - 1;

          if(i > 0 && nums[i] == nums[i-1]) {
              continue;
          }

          int sum;
          while (l < r) {
              sum = currentElement + nums[l] + nums[r];
              if(sum < 0) l++;
              else if(sum > 0) r--;
              else {
                  result.add(Arrays.asList(currentElement, nums[l++], nums[r]));

                  while (nums[l] == nums[l - 1] && l < r) l += 1;
              }
          }
      }

      return result;
  }
}

File: py/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        a=nums
        a.sort()
        i=0
       
        l=[]
        ss=set()
        while i<len(nums)-2:
            j=i+1
            k=len(nums)-1
            temp=0-a[i]
            while j<k:
                if a[j]+a[k]==temp:
                    if (a[i],a[j],a[k]) not in ss:
                        l.append([a[i],a[j],a[k]])
                        ss.add((a[i],a[j],a[k]))
                    k=k-1
                    j=j+1
                elif  a[j]+a[k]>temp:
                    
                    k=k-1
                else:
                    
                    j=j+1
            i=i+1
        return l
                    
                
